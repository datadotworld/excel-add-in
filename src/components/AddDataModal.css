/*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */
.add-data-modal {
  top: 50px;
}

.selection {
  display: flex;
  border: 1px solid #a3afbf;
  justify-content: space-between;
  align-items: center;
}

.selection span {
  margin: 0;
  padding-left: 6px;
  display: flex;
  align-items: center;
  font-weight: normal;
}

.selection span::before {
  content: '';
  display: inline-block;
  width: 1em;
  height: 1em;
  border-radius: 1em;
  border: 1px solid white;
  box-shadow: 0 0 0 0.05em grey;
  margin-right: 0.75em;
}

.selection input[type=radio] {
  display: none;
}

.selection input[type=radio]:checked + span::before {
  border: 4.5px solid #328df3;
  background: white;
  box-shadow: 0 0 0 0.05em grey;
}

.selection .radio-button {
  display: flex;
  padding: 12px;
  align-items: center;
}

.selection .selection-info {
  display: flex;
  padding: 12px;
  color: #969696;
  font-size: 11px;
}

.select-sheet {
  background-color: aliceblue;
}

.full-screen-modal .center-block .selection-form {
  padding: 0;
}
