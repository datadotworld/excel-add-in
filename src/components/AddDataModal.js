/*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { 
  Button,
  ControlLabel,
  FormControl,
  FormGroup,
  Grid,
  HelpBlock,
  InputGroup,
  Row
} from 'react-bootstrap';

import analytics from '../analytics';

import Icon from './icons/Icon';
import WarningModal from './WarningModal';
import './AddDataModal.css';

const filenameRegex = /^[^/]+$/;
const MAX_FILENAME_LENGTH = 128;

const getFilenameWithoutExtension = function (filename) {
  const dotPos = filename.lastIndexOf('.');
  return dotPos > -1 ? filename.slice(0, dotPos) : filename;
}

const selectSheetState = function (options) {
  const binding = options.binding;
  // If editing a binding set the state from the binding
  if (binding) {
    const maxRows = 1048576;
    const maxColumns = 150;

    // If it's a sheet binding set to true
    if (binding.rowCount === maxRows && binding.columnCount === maxColumns) {
      return true;
    }
  }

  // False when creating a new binding or editing a selection binding
  return false;
}

class AddDataModal extends Component {
  static propTypes = {
    close: PropTypes.func,
    createBinding: PropTypes.func,
    doesFileExist: PropTypes.func.isRequired,
    excelApiSupported: PropTypes.bool,
    options: PropTypes.object,
    range: PropTypes.object,
    refreshLinkedDataset: PropTypes.func,
    sync: PropTypes.func,
    updateBinding: PropTypes.func
  }

  static defaultProps = {
    options: {}
  }

  state = {
    isSubmitting: false,
    name: this.props.options.filename ? getFilenameWithoutExtension(this.props.options.filename) : '',
    showWarningModal: false,
    selectSheet: selectSheetState(this.props.options)
  }

  isFormValid = () => {
    const { name } = this.state;
    const { excelApiSupported, range } = this.props;
    if ((excelApiSupported && !range) || !name) {
      return false;
    }
    return name.match(filenameRegex) && name.length < MAX_FILENAME_LENGTH;
  }

  submitBinding = () => {
    this.closeModal();
    const {
      close,
      createBinding,
      options,
      refreshLinkedDataset,
      sync,
      updateBinding,
      range
    } = this.props;
    const { name, selectSheet } = this.state;

    const allRows = 1048576;
    const selection = {
      name: `${name}.csv`,
      sheetId: range.worksheet.id,
      range: selectSheet ? `A1:ET${allRows}` : range.address
    };

    if (options.binding) {
      updateBinding(options.binding, selection)
        .then(sync)
        .then(refreshLinkedDataset)
        .then(close);
    } else {
      createBinding(selection).then((binding) => {
        // Binding has been created, but the file does not exist yet, sync the file
        sync(binding).then(refreshLinkedDataset).then(close);
      })
    }
  }

  submit = (event) => {
    analytics.track('exceladdin.add_data.submit.click');
    event.preventDefault();

    if (this.props.options.filename || this.props.doesFileExist(`${this.getFilename(this.state.name)}.csv`)) {
      // Show warning modal
      this.setState({ showWarningModal: true });
    } else {
      this.submitBinding();
    }
  }

  closeModal = () => {
    this.setState({ showWarningModal: false });
  }

  /**
   * Peels off a trailing .csv if the user added it
   */
  getFilename = () => {
    const index = this.state.name.lastIndexOf('.csv');
    if (index >= 0 && index === this.state.name.length - 4) {
      return getFilenameWithoutExtension(this.state.name);
    }
    return this.state.name;
  }

  closeClicked = () => {
    analytics.track('exceladdin.add_data.close.click');
    this.props.close();
  }

  cancelClicked = () => {
    analytics.track('exceladdin.add_data.cancel.click');
    this.props.close();
  }

  getSheetName = (range) => {
    if (range) {
      const address = range.address;
      // Extract sheet name and trim quote marks which are added when the name has a space
      const sheet = address.substring(0, address.indexOf('!')).replace(/'/g, '');
      return sheet;
    }

    return ''
  }

  changeSelection = (event) => {
    const { value } = event.target

    if (value === 'selection') {
      this.setState({ selectSheet: false})
    } else if (value === 'sheet') {
      this.setState({ selectSheet: true });
    }
  }

  getSelectionText(range) {
    if (range) {
      // Return number of rows and columns in selection
      const rowCount = range.rowCount;
      const columnCount = range.columnCount;
      const rowText = rowCount === 1 ? 'Row' : 'Rows';
      const columnText = columnCount === 1 ? 'Column' : 'Columns';

      return `(${rowCount} ${rowText} x ${columnCount} ${columnText})`;
    }

    // Range no yet loaded, return placeholder text
    return '(Rows x Columns)';
  }

  render () {
    const { name, selectSheet } = this.state;
    const { excelApiSupported, options, range } = this.props;

    let validState, displayName, selection;

    if (name) {
      validState = this.isFormValid() ? 'success' : 'warning'
      displayName = this.getFilename();
    }

    if (selectSheet) {
      selection = range ? this.getSheetName(range) : '';
    } else {
      selection = range ? range.address : '';
    }

    return (
      <Grid className='add-data-modal full-screen-modal show'>
        <Row className='center-block header'>
          <div className='title'>
            Add data <Icon icon='close' className='close-button' onClick={this.closeClicked} />
          </div>
        </Row>
        <Row className='center-block'>
          <form onSubmit={this.submit}>
            {excelApiSupported &&
            <div className='selection-form' onClick={this.changeSelection}>
              <div className='selection'>
                <label className='radio-button'>
                  <input
                    type='radio'
                    name='selection'
                    value='selection'
                    checked={!this.state.selectSheet}
                    readOnly
                  />
                  <span>Selection</span>
                </label>
                <div className='selection-info'>
                  { this.getSelectionText(range) }
                </div>
              </div>
              <div className='selection'>
                <label className='radio-button'>
                  <input
                    type='radio'
                    name='selection'
                    value='sheet'
                    checked={this.state.selectSheet}
                    readOnly
                  />
                  <span>Sheet</span>
                </label>
                <div className='selection-info'>
                  {`(${this.getSheetName(range)})`}
                </div>
              </div>
              <HelpBlock>
                {
                  selectSheet ?
                  'Select the sheet you\'d like to add, it will be refelected here.' :
                  'Select the area to bind in the worksheet, it will be reflected here.'
                }
              </HelpBlock>
            </div>}
            {!excelApiSupported && <div>
              <ControlLabel>Dataset range</ControlLabel>
              <HelpBlock>
                Select the area to bind in the worksheet.
              </HelpBlock>
            </div>}
            <FormGroup validationState={validState}>
              <ControlLabel>Name</ControlLabel>
              <InputGroup>
                <FormControl
                  onChange={(event) => this.setState({name: event.target.value})}
                  value={name}
                  disabled={!!options.filename}
                  type='text' />
              </InputGroup>
              <HelpBlock>
                Must not include slashes, your file will be named <strong>{displayName}.csv</strong>
                <div className='titleLimit'>max. 128</div>
              </HelpBlock>
            </FormGroup>
            <div className='button-group'>
              <Button onClick={this.cancelClicked}>Cancel</Button>
              <Button
                type='submit'
                disabled={this.state.isSubmitting || validState !== 'success'}
                bsStyle='primary'>OK</Button>
            </div>
          </form>
        </Row>
        <WarningModal
          show={this.state.showWarningModal}
          cancelHandler={this.closeModal}
          successHandler={this.submitBinding}
          analyticsLocation='exceladdin.add_data'>
          <div><strong>"{displayName}.csv" already exists on data.world.  Do you want to replace it?</strong></div>
          <div>Replacing it will overwrite the file on data.world with the contents from {excelApiSupported ? selection : 'the selected cell range'}</div>
        </WarningModal>
      </Grid>);
  }
}

export default AddDataModal;