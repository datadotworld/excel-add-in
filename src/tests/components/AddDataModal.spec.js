/*
 * Copyright 2017 data.world, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * This product includes software developed at
 * data.world, Inc. (http://data.world/).
 */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';

import AddDataModal from '../../components/AddDataModal';

it('renders modal', () => {
  expect(renderer.create(<AddDataModal />).toJSON()).toMatchSnapshot()
});

it('renders modal - with range', () => {
  expect(renderer.create(
    <AddDataModal
      range={{
        rowCount: 4,
        columnCount: 2,
        address: 'Sheet1!A2:B5'
      }}
      excelApiSupported
    />
  ).toJSON()).toMatchSnapshot()
});

it('renders modal - with options', () => {
  const options = {
    filename: 'test1.csv'
  };
  expect(renderer.create(
    <AddDataModal
      range={{
        rowCount: 4,
        columnCount: 2,
        address: 'Sheet1!A2:B5'
      }}
      options={options}
      excelApiSupported
    />
  ).toJSON()).toMatchSnapshot()
});

it ('removes trailing .csv from filenames', () => {
  const modal = new AddDataModal({options: {}});
  modal.state.name = 'testing.csv'
  expect(modal.getFilename()).toBe('testing');
});

it ('does not impact filenames without trailing .csv', () => {
  const modal = new AddDataModal({options: {}});
  modal.state.name = 'testingwithoutextension'
  expect(modal.getFilename()).toBe('testingwithoutextension');
});