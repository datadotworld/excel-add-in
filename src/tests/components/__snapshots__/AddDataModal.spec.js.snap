// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders modal - with options 1`] = `
<div
  className="add-data-modal modal show container"
>
  <div
    className="center-block header row"
  >
    <div
      className="title"
    >
      Add data 
      <span
        className="svg-icon close-button"
        onClick={[Function]}
      >
        <svg
          className="close-icon"
          height="15"
          viewBox="0 0 16 15"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            className="stroke stroke-medium"
            d="M15.222 14.222L1 0M1 14.222L15.222 0"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    className="center-block row"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Dataset range
        </label>
        <span
          className="input-group"
        >
          <input
            className="form-control"
            disabled={true}
            id={undefined}
            type="text"
            value="Sheet1!A2:B5"
          />
        </span>
        <span
          className="help-block"
        >
          Select the area to bind in the worksheet, it will be reflected here.
        </span>
      </div>
      <div
        className="form-group has-success"
      >
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Name
        </label>
        <span
          className="input-group"
        >
          <input
            className="form-control"
            disabled={true}
            id={undefined}
            onChange={[Function]}
            type="text"
            value="test1"
          />
        </span>
        <span
          className="help-block"
        >
          Must not include slashes, your file will be named 
          <strong>
            test1
            .csv
          </strong>
          <div
            className="titleLimit"
          >
            max. 128
          </div>
        </span>
      </div>
      <div
        className="button-group"
      >
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
        <button
          className="btn btn-primary"
          disabled={false}
          type="submit"
        >
          Add selection
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`renders modal - with range 1`] = `
<div
  className="add-data-modal modal show container"
>
  <div
    className="center-block header row"
  >
    <div
      className="title"
    >
      Add data 
      <span
        className="svg-icon close-button"
        onClick={[Function]}
      >
        <svg
          className="close-icon"
          height="15"
          viewBox="0 0 16 15"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            className="stroke stroke-medium"
            d="M15.222 14.222L1 0M1 14.222L15.222 0"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    className="center-block row"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Dataset range
        </label>
        <span
          className="input-group"
        >
          <input
            className="form-control"
            disabled={true}
            id={undefined}
            type="text"
            value="Sheet1!A2:B5"
          />
        </span>
        <span
          className="help-block"
        >
          Select the area to bind in the worksheet, it will be reflected here.
        </span>
      </div>
      <div
        className="form-group"
      >
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Name
        </label>
        <span
          className="input-group"
        >
          <input
            className="form-control"
            disabled={false}
            id={undefined}
            onChange={[Function]}
            type="text"
            value=""
          />
        </span>
        <span
          className="help-block"
        >
          Must not include slashes, your file will be named 
          <strong>
            .csv
          </strong>
          <div
            className="titleLimit"
          >
            max. 128
          </div>
        </span>
      </div>
      <div
        className="button-group"
      >
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
        <button
          className="btn btn-primary"
          disabled={true}
          type="submit"
        >
          Add selection
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`renders modal 1`] = `
<div
  className="add-data-modal modal show container"
>
  <div
    className="center-block header row"
  >
    <div
      className="title"
    >
      Add data 
      <span
        className="svg-icon close-button"
        onClick={[Function]}
      >
        <svg
          className="close-icon"
          height="15"
          viewBox="0 0 16 15"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            className="stroke stroke-medium"
            d="M15.222 14.222L1 0M1 14.222L15.222 0"
          />
        </svg>
      </span>
    </div>
  </div>
  <div
    className="center-block row"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Dataset range
        </label>
        <span
          className="input-group"
        >
          <input
            className="form-control"
            disabled={true}
            id={undefined}
            type="text"
            value="Unavailable in Excel 2013"
          />
        </span>
        <span
          className="help-block"
        >
          Select the area to bind in the worksheet, it will be reflected here.
        </span>
      </div>
      <div
        className="form-group"
      >
        <label
          className="control-label"
          htmlFor={undefined}
        >
          Name
        </label>
        <span
          className="input-group"
        >
          <input
            className="form-control"
            disabled={false}
            id={undefined}
            onChange={[Function]}
            type="text"
            value=""
          />
        </span>
        <span
          className="help-block"
        >
          Must not include slashes, your file will be named 
          <strong>
            .csv
          </strong>
          <div
            className="titleLimit"
          >
            max. 128
          </div>
        </span>
      </div>
      <div
        className="button-group"
      >
        <button
          className="btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
        <button
          className="btn btn-primary"
          disabled={true}
          type="submit"
        >
          Add selection
        </button>
      </div>
    </form>
  </div>
</div>
`;
